# 01 Backwards and Forwards Models

The final bit for today is looking at ensembling backwards and forwards models. 

The technique can be boiled down as simply reversing the entire sentence. We train two models, one forwards and one backwards, and ensemble the two of them. Both the language model and the downstream'd model have this augmentation applied.

We need to make two modifications to get this to work, a transform to reverse and a modified `learner` to enable us to get the pretrained langauge model:

In [ ]:
!pip install fastai2

In [ ]:
from fastai2.text.all import *

In [ ]:
path = untar_data(URLs.IMDB_SAMPLE)

In [ ]:
df = pd.read_csv(path/'texts.csv')

We reverse each individual text inside of the batch:

In [ ]:
def reverse_text(x): return torch.stack([a.flip(0) for a in x])

Let's build two LM DataLoaders to see this in action, one with reverse and one without:

First the regular (forwards) `DataBlock` and `DataLoaders`:

In [ ]:
lm_block = TextBlock.from_df('text', is_lm=True, res_col_name='tok_text')
dblock = DataBlock(blocks=lm_block,
                   get_x=ColReader('tok_text'),
                   splitter=RandomSplitter(0.1, seed=42))

In [ ]:
dls = dblock.dataloaders(df, bs=64, seq_len=72, num_workers=4)

Then our new backwards one, with an added `batch_tfm` of our `reverse_text` function:

In [ ]:
lm_block = TextBlock.from_df('text', is_lm=True, res_col_name='tok_text')
dblock = DataBlock(blocks=lm_block,
                   get_x=ColReader('tok_text'),
                   splitter=RandomSplitter(0.1, seed=42),
                   batch_tfms=reverse_text)

In [ ]:
dls_r = dblock.dataloaders(df, bs=64, seq_len=72, num_workers=4)

Let's take a look at our backwards `DataLoader` to make sure it is working:

In [ ]:
dls_r.show_batch()

So how do we know? Let's look at the first batch in our validation sets:

In [ ]:
b_f = next(iter(dls[1]))
b_b = next(iter(dls_r[1]))

In [ ]:
b_f[0]

LMTensorText([[   2,    8,   28,  ...,    0,   12,  211],
        [ 637,   31,   33,  ...,   41,  286,   18],
        [  15,  345, 3432,  ...,  114, 2502,   12],
        ...,
        [ 246,    0,   11,  ...,   19,   85,   44],
        [ 500,   37,   43,  ..., 3801,  798,   66],
        [3014,   10,    8,  ...,   24,  124,   54]], device='cuda:0')

In [ ]:
b_b[0]

LMTensorText([[ 211,   12,    0,  ...,   28,    8,    2],
        [  18,  286,   41,  ...,   33,   31,  637],
        [  12, 2502,  114,  ..., 3432,  345,   15],
        ...,
        [  44,   85,   19,  ...,   11,    0,  246],
        [  66,  798, 3801,  ...,   43,   37,  500],
        [  54,  124,   24,  ...,    8,   10, 3014]], device='cuda:0')

We can see that each word is simply 100% reversed. There are some other augmentation techniques too, such as translation that we may look at later

The last thing we need is a way to enable us to grab the backwards language model inside of `language_model_learner`:

In [ ]:
def _get_text_vocab(dls):
    vocab = dls.vocab
    if isinstance(vocab, L): vocab = vocab[0]
    return vocab

from fastai2.text.models.core import _model_meta

def language_model_learner(dls, arch, config=None, drop_mult=1., backwards=True, pretrained=True, pretrained_fnames=None, **kwargs):
    "Create a `Learner` with a language model from `dls` and `arch`."
    vocab = _get_text_vocab(dls)
    model = get_language_model(arch, len(vocab), config=config, drop_mult=drop_mult)
    meta = _model_meta[arch]
    learn = LMLearner(dls, model, loss_func=CrossEntropyLossFlat(), splitter=meta['split_lm'], **kwargs)
    #TODO: add backard
    url = 'url_bwd' if backwards else 'url'
    if pretrained or pretrained_fnames:
        if pretrained_fnames is not None:
            fnames = [learn.path/learn.model_dir/f'{fn}.{ext}' for fn,ext in zip(pretrained_fnames, ['pth', 'pkl'])]
        else:
            if url not in meta:
                warn("There are no pretrained weights for that architecture yet!")
                return learn
            model_path = untar_data(meta[url] , c_key='model')
            fnames = [list(model_path.glob(f'*.{ext}'))[0] for ext in ['pth', 'pkl']]
        learn = learn.load_pretrained(*fnames)
    return learn

Now this is still in the in-progress stage at this moment, it'll be more in-house once it's finished and I'll update this notebook accordingly

## Training

So for training, this will be an ensemble of two models, a forwards and a backwards model. We'll train all the models all at once and save them away into an array:

In [ ]:
def train_models(models:list):
    names = ['fwd', 'bwd']
    for i, model in enumerate(models):
        lr = 1e-2
        lr *= model.dls.bs/48
        model = model.to_fp16()
        model.fine_tune(5)
        model.save_encoder(f'{names[i]}_fine_tuned_enc')
    return models

In [ ]:
fwd = language_model_learner(dls, AWD_LSTM, backwards=False, metrics=[accuracy, Perplexity()])
bwd = language_model_learner(dls_r, AWD_LSTM, backwards=True, metrics=[accuracy, Perplexity()])

In [ ]:
spp = train_models([fwd, bwd])

Now we just repeat what we did last time!

In [ ]:
blocks = (TextBlock.from_df('text', res_col_name='tok_text', seq_len=fwd.dls.seq_len,
                            vocab=fwd.dls.vocab), CategoryBlock())

imdb_class = DataBlock(blocks=blocks,
                       get_x=ColReader('tok_text'),
                       get_y=ColReader('label'),
                       splitter=ColSplitter(col='is_valid'))
fwd_dls = imdb_class.dataloaders(df, bs=32)

In [ ]:
def reverse_text(x:TensorText): return torch.stack([a.flip(0) for a in x])

In [ ]:
blocks = (TextBlock.from_df('text', res_col_name='tok_text', seq_len=bwd.dls.seq_len,
                            vocab=bwd.dls.vocab), CategoryBlock())

imdb_class = DataBlock(blocks=blocks,
                       get_x=ColReader('tok_text'),
                       get_y=ColReader('label'),
                       splitter=ColSplitter(col='is_valid'),
                       batch_tfms=reverse_text)
bwd_dls = imdb_class.dataloaders(df, bs=32)

In [ ]:
bwd_dls.show_batch()

In [ ]:
fwd = text_classifier_learner(fwd_dls, AWD_LSTM, backwards=False, metrics=[accuracy])
bwd = text_classifier_learner(bwd_dls, AWD_LSTM, backwards=True, metrics=[accuracy])

In [ ]:
fwd.load_encoder('fwd_fine_tuned_enc');
bwd.load_encoder('bwd_fine_tuned_enc');

In [ ]:
def train_class(models:list):
    for i, model in enumerate(models):
        lr = 0.04365158379077912
        adj = 2.6**4
        model.fit_one_cycle(1, lr)
        model.freeze_to(-2)
        model.fit_one_cycle(1, slice(lr/adj, lr))
        model.freeze_to(-3)
        lr /= 2
        model.fit_one_cycle(1, slice(lr/adj, lr))
        model.unfreeze()
        lr /= 5
        model.fit_one_cycle(2, slice(lr/adj, lr))
    return models

In [ ]:
models = [fwd,bwd]

In [ ]:
models = train_class(models)

Now that we have our model, let's ensemble!

In [ ]:
fwd_p, y = models[0].get_preds()
bwd_p, _ = models[1].get_preds()

In [ ]:
p = (fwd_p + bwd_p) / 2

In [ ]:
accuracy(p, y)

tensor(0.8300)

Let's compare that to both our others:

In [ ]:
accuracy(fwd_p, y)

tensor(0.8100)

In [ ]:
accuracy(bwd_p, y)

tensor(0.7950)

|     Model Type    | Accuracy |
|:-----------------:|:--------:|
| Forwards (Normal) |  81.00%  |
|     Backwards     |  79.50%  |
|      Ensemble     |  83.00%  |
|<img width=200/>|<img width=200/>|

We can see that even though they were well below 83%, combined they boosted up quite high!